pkg/preemptiontoleration/preemption_toleration.go

```diff
@@ -0,0 +1,429 @@
+	nonViolatingCandidates := newCandidateList(numCandidates)
+	violatingCandidates := newCandidateList(numCandidates)
+	parallelCtx, cancel := context.WithCancel(ctx)
+	nodeStatuses := make(framework.NodeToStatusMap)
+	var statusesLock sync.Mutex
+	now := pl.clock.Now()
+	checkNode := func(i int) {
+		nodeInfoCopy := potentialNodes[(int(offset)+i)%len(potentialNodes)].Clone()
+		stateCopy := state.Clone()
+		pods, numPDBViolations, status := pl.selectVictimsOnNode(ctx, fh, stateCopy, pod, nodeInfoCopy, pdbs, now)
+		if status.IsSuccess() {
+			victims := extenderv1.Victims{
+				Pods:             pods,
+				NumPDBViolations: int64(numPDBViolations),
+			}
+			c := &candidate{
+				victims: &victims,
+				name:    nodeInfoCopy.Node().Name,
+			}
+			if numPDBViolations == 0 {
+				nonViolatingCandidates.add(c)
+			} else {
+				violatingCandidates.add(c)
+			}
+			nvcSize, vcSize := nonViolatingCandidates.size(), violatingCandidates.size()
+			if nvcSize > 0 && nvcSize+vcSize >= numCandidates {
+				cancel()
+			}
+		} else {
+			statusesLock.Lock()
+			nodeStatuses[nodeInfoCopy.Node().Name] = status
+			statusesLock.Unlock()
+		}
+	}
+	fh.Parallelizer().Until(parallelCtx, len(potentialNodes), checkNode)
+	return append(nonViolatingCandidates.get(), violatingCandidates.get()...), nodeStatuses
+}
+
+// selectVictimsOnNode finds minimum set of pods on the given node that should
+// be preempted in order to make enough room for "preemptor" to be scheduled.
+// The algorithm is almost identical to DefaultPreemption plugin's one.
+// The only difference is that it takes PreemptionToleration annotations in
+// PriorityClass resources into account for selecting victim pods.
+func (pl *PreemptionToleration) selectVictimsOnNode(
+	ctx context.Context,
+	fh framework.Handle,
+	state *framework.CycleState,
+	preemptor *v1.Pod,
+	nodeInfo *framework.NodeInfo,
+	pdbs []*policy.PodDisruptionBudget,
+	now time.Time,
+) ([]*v1.Pod, int, *framework.Status) {
+	var potentialVictims []*framework.PodInfo
+	removePod := func(rpi *framework.PodInfo) error {
+		if err := nodeInfo.RemovePod(rpi.Pod); err != nil {
+			return err
+		}
+		status := fh.RunPreFilterExtensionRemovePod(ctx, state, preemptor, rpi, nodeInfo)
+		if !status.IsSuccess() {
+			return status.AsError()
+		}
+		return nil
+	}
+	addPod := func(api *framework.PodInfo) error {
+		nodeInfo.AddPodInfo(api)
+		status := fh.RunPreFilterExtensionAddPod(ctx, state, preemptor, api, nodeInfo)
+		if !status.IsSuccess() {
+			return status.AsError()
+		}
+		return nil
+	}
+
+	// As the first step, remove all lower priority pods that can't tolerate preemption by the preemptor
+	// from the node and check if the given pod can be scheduled.
+	podPriority := corev1helpers.PodPriority(preemptor)
+	for _, pi := range nodeInfo.Pods {
+		if corev1helpers.PodPriority(pi.Pod) >= podPriority {
+			continue
+		}
+
```

Comment:
Add a comment here:

```
For a pod with lower priority, check if it can be exempted from the preemption.
```

BTW: I think "exempted" reads better than "tolerated", can we change the wordings here?

```
exempted, err := ExemptedFromPreemption(pi.Pod, preemptor, pl.priorityClassLister, now)
```

Reply:
fixed in https://github.com/kubernetes-sigs/scheduler-plugins/pull/224/commits/aa772775201193d59e5cbd12502ea31d922df6c7